### PR TITLE
Introducing services!

### PR DESCRIPTION
**Pull request by edgardz**
Jira Ticket ID: <JIRA Ticket ID> 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Code style update
- [ ] Refactor (refactoring or adding test which isn't a fix or add a feature)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**Did you test your solution?**

- [ ] I lightly tested it in one browser
- [x] I deeply tested it in several browsers
- [ ] I wrote tests around it (unit tests, integration tests, E2E tests)

## Problem Description

Sometimes we get into performance problems due to the number of event listeners present in a project. For instance, when we have several components in a list that listen to the resize event, or when we are developing a game and end up with several requestAnimationFrames scattered throughout the project, it becomes hard to manage and performance suffers.

## Solution Description

Introducing services!

```
src/
    services/
        orientation.js
        resize.js
        visibility.js
        ...
```

This is very useful. If you ever worked with Angular this will be straightforward to you.  
A service is typically a class with a narrow, well-defined purpose. It should do something specific and do it well.  
Services are [singleton](https://refactoring.guru/design-patterns/singleton) class instances that abstract, and prevent, the need of adding several event listeners for the same browser events, like resize, requestAnimationFrame, etc. We should use services for events that are prone to be listened to by many components.

## Side Effects, Risks, Impact

- [x] N/A

**Aditional comments:**

Services like these (and many others) were successfully used on several Jam3 projects already, like "Spotify The Weeknd" and "Facebook Connect".

### CODE REVIEW

File: src/services/orientation.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+type OrientationListener = (e?: Event) => void;
+
+class OrientationService {
+  listeners: OrientationListener[] = [];
+
+  onOrientation = (e: Event) => {
+    this.listeners.map((listener) => listener(e));
```
Comment (DonghyukJacobJang) [inline src/services/orientation.ts:7]:
```suggestion
    this.listeners.forEach((listener) => listener(e));
```

File: src/services/pointer-move.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+type PointerMoveListener = (x?: number, y?: number, e?: MouseEvent | TouchEvent) => void;
+
+class PointerMoveService {
+  listeners: PointerMoveListener[] = [];
+
+  onMove = (e: MouseEvent | TouchEvent) => {
+    const x = (e as TouchEvent).targetTouches?.[0]?.clientX || (e as MouseEvent).clientX;
+    const y = (e as TouchEvent).targetTouches?.[0]?.clientY || (e as MouseEvent).clientY;
+    this.listeners.map((listener) => listener(x, y, e));
```
Comment (DonghyukJacobJang) [inline src/services/pointer-move.ts:11]:
```suggestion
    this.listeners.forEach((listener) => listener(x, y, e));
```

File: src/services/raf.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+class RequestAnimationFrameService {
+  listeners: RequestAnimationFrameListener[] = [];
+  frameId = 0;
+  elapsed = 0;
+
+  onFrame = () => {
+    const now = Date.now();
+    const delta = now - this.elapsed;
+    this.elapsed = now;
+    this.listeners.map((listener) => listener(delta));
```
Comment (DonghyukJacobJang) [inline src/services/raf.ts:12]:
```suggestion
    this.listeners.forEach((listener) => listener(delta));
```

File: src/services/resize.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+type ResizeListener = (e?: Event | UIEvent) => void;
+
+class ResizeService {
+  listeners: ResizeListener[] = [];
+
+  onResize = (e: Event | UIEvent) => {
+    setTimeout(
+      () => {
+        this.listeners.map((listener) => listener(e));
```
Comment (DonghyukJacobJang) [inline src/services/resize.ts:11]:
```suggestion
        this.listeners.forEach((listener) => listener(e));
```

File: src/services/scroll.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+type ScrollListener = (e?: Event) => void;
+
+class ScrollService {
+  listeners: ScrollListener[] = [];
+
+  onScroll = (e: Event) => {
+    this.listeners.map((listener) => listener(e));
```
Comment (DonghyukJacobJang) [inline src/services/scroll.ts:7]:
```suggestion
    this.listeners.forEach((listener) => listener(e));
```

File: src/services/visibility.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+type VisibilityListener = (e?: Event) => void;
+
+class VisibilityService {
+  listeners: VisibilityListener[] = [];
+
+  onVisibility = (e: Event) => {
+    this.listeners.map((listener) => listener(e));
```
Comment (DonghyukJacobJang) [inline src/services/visibility.ts:7]:
```suggestion
    this.listeners.forEach((listener) => listener(e));
```

Comment (edgardz) [reply src/services/visibility.ts]: All done. good catch.

**Review by DonghyukJacobJang — APPROVED** [review]

**Pull request merged**